cedar-policy-core/src/ast/request.rs

```diff
@@ -164,13 +165,15 @@ impl Context {
     }
 
     /// Create a `Context` from a `RestrictedExpr`, which must be a `Record`
+    /// INVARIANT: `from_expr` must only be called with an `expr` that is a `Record`
     pub fn from_expr(expr: RestrictedExpr) -> Self {
         debug_assert!(matches!(expr.expr_kind(), ExprKind::Record { .. }));
         Self { context: expr }
     }
 
     /// Create a `Context` from a map of key to `RestrictedExpr`, or a Vec of
     /// `(key, RestrictedExpr)` pairs, or any other iterator of `(key, RestrictedExpr)` pairs
+    // INVARIANT: always constructs a record
     pub fn from_pairs(pairs: impl IntoIterator<Item = (SmolStr, RestrictedExpr)>) -> Self {
```

Comment:
in one of these, the `INVARIANT` is in the doc string, in the other it is not

Reply:
I put it in a docstring when it's something the caller needs to consider

Reply:
I think the second isn't an invariant on the functions, it's just stating that the function maintains the datatype invariant.